Feature/kas 2632 pub proof  delete piece

mdv-professional opened the pull request:
[Jira](https://kanselarij.atlassian.net/browse/KAS-2632)

**Notes**
N1. Ik heb gebruik gemaakt van Row-objecten. Ik ben zelf nog steeds meer voorstander van de `computed` helper, omdat het aanmaken van Row-objecten complex is door async + service injection. Ze zijn wel nuttig om de file-property los te koppelen van het piece-object: tijdens een delete wordt er anders een error gelogd i.v.m. een binding (niet zichtbaar in de UI). Ik vermoed dat deze ermee te maken heeft dat de relatie piece <-> file niet meer bestaat. Als ik van file een property van het Row-object maak treedt ze niet meer op. De bindings in de template deze nog wel verwachten. De fout zit ook in vertalingen.
N2. @brenner-company Gewoon omdat het al even geleden is, disabled buttons in de dropdown zouden goed zijn.

**Extra**
X1. (gedaan) Aan de delete van vertalingen heb ik ook de check voor OVRB toegevoegd. Dit staat niet in het issue, maar was besproken met Michael, en ik denk dat ik de change ergens kwijt ben gespeeld. UPDATE: misschien heb ik me hier vergist.
X2. (niet gedaan) Ik had graag ook Row-objecten gemaak voor de pieces van vertalingen, om dat error logging issue op te lossen.

Commit:
KAS-2632 sorting issues          | delete piece [pub proof]

Commit:
KAS-2632 PieceRow and Model          | delete piece [pub proof]

Commit:
KAS-2632 refactor (busy)          | delete piece [pub proof]

Commit:
KAS-2632 OVRB only          | delete piece [pub proof]

Commit:
KAS-2632 restore sort          | delete piece [pub proof]

Commit:
KAS-2632 small cleanup          | delete piece [pub proof]

brenner-company commented:
@mdv-professional Voor die disabled state ben ik nu even aan het kijken. Ik hou je op de hoogte 👍

ghost commented:
OK, ga nu die nieuwe styling bekijken.
Ik heb ook opgemerkt dat de delete-check nog fout zit.

Commit:
KAS-2632 no delete received                       | delete piece [pub proof]

Commit:
KAS-2632 clarification                            | delete piece [pub proof]

MikiDi committed:
Merge branch 'development' into feature/KAS-2632-pub-proof--delete-piece

MikiDi commented:
Gesynced & merge-conflicts opgelost na mergen van #1030 en #1031

ValenberghsSven commented on app/pods/publications/publication/proofs/documents/controller.js:
```diff
@@ -160,6 +217,27 @@ export default class PublicationsPublicationProofsDocumentsController extends Co
+    const destroyFile = file.destroyRecord();
+    const destroyDocumentContainer = documentContainer.destroyRecord();
+
+    yield Promise.all([destroyPiece, destroyFile, destroyDocumentContainer]);
```
Je zegt een probleem te hebben met errors, een reden hiervoor kan zijn dat we normaal eerst de file verwijderen, dan de piece, en dan de container. De volgorde van verwijderen is hier anders.
edit: lokaal getest en is niet voldoende (getest bij vertalingen). Die fileTypePill is de oorzaak

ValenberghsSven commented on app/pods/publications/publication/proofs/documents/controller.js:
```diff
@@ -1,12 +1,65 @@
+    model.decisions = decisions;
+    return model;
+  }
+}
```
Het PieceRow object gebruiken kan ik volgen, maar is het nodig om ook het model op deze manier te overschrijven?
Kunnen die paar regels niet gewoon bij in de `route.js` ?

Ik zie daar de meerwaarde niet van, en het kan verwarrend gaan worden als we op meerdere plaatsen het veel gebruikte `model` gaan overschrijven met een custom object met zelfde naam.

ValenberghsSven commented on app/pods/publications/publication/proofs/documents/controller.js:
```diff
@@ -1,12 +1,65 @@
+    const row = new PieceRow();
+    row.piece = piece;
+    row.file = await piece.file;
+    row.publicationSubcase = publicationSubcase;
```
publicationSubcase is voor elke rij hetzelfde normaal (1 per publicatie) ? Dan kan dat gewoon uit model/controller komen zou ik zeggen.

ValenberghsSven commented on app/pods/publications/publication/proofs/documents/controller.js:
```diff
@@ -1,12 +1,65 @@
+    row.publicationSubcase = publicationSubcase;
+    // avoid awaiting in getter
+    row.requestActivitiesUsedBy = await piece.requestActivitiesUsedBy;
+    row.currentSession = currentSession;
```
currentSession zou iets "globaal" moeten zijn, en zeker niet op rij niveau.
Dit komt gewoon rechtreeks van de controller als je dat wil gebruiken

ValenberghsSven commented on app/pods/publications/publication/proofs/documents/route.js:
```diff
@@ -56,10 +61,7 @@ export default class PublicationsPublicationProofsDocumentsRoute extends Route {
-      pieces: pieces,
-      decisions: decisions,
-    };
+    return Model.create(pieces, decisions, this.publicationSubcase, this.currentSession);
```
Zoals opgemerkt in de controller, kan de logica van die Model.create niet gewoon hier?
Dan heb je nog steeds je pieceRows, maar gaan we geen custom model object maken

ValenberghsSven commented on app/pods/publications/publication/proofs/documents/controller.js:
```diff
@@ -1,12 +1,65 @@
+  static async create(piece, publicationSubcase, currentSession) {
+    const row = new PieceRow();
+    row.piece = piece;
+    row.file = await piece.file;
+    row.publicationSubcase = publicationSubcase;
+    // avoid awaiting in getter
+    row.requestActivitiesUsedBy = await piece.requestActivitiesUsedBy;
```
Die awaits van file en requestActivities, zijn deze niet al resolved door de includes die gebeurd zijn in de route logica?

erikap replied:
We hebben met @mdv-professional afgesproken om in het geval van een `include` de `await` toch expliciet te zetten altijd ook al weet je dat die promise onmiddellijk zal resolven. Als we die `include` in de toekomst zouden weghalen, blijft de code dan toch werken.

ValenberghsSven reviewed (changes requested):
Betreffende die pieceRow object.
Ik ben er op zich niet tegen, Het werken met een rij met allerlei informatie wordt dan iets makkelijker. Het kan werken met, het kan werken zonder.
Maar als je mijn opmerkingen verwerkt heb je enkel `piece`, `file`, en `requestActivitiesUsedBy` over in je object. Dan is het gebruik van een object nog verantwoord, maar op veel plaatsen doen we nu iets gelijkaardig zonder object.

Het Model object ben ik wel tegen. Ik zie daar niet direct een duidelijke reden voor, dat kan volgens mij los in de route gebeuren wat daarin zit.

Het probleem van de error bij verwijderen ligt eerder aan de `FileTypePill`, die kan er niet deftig mee om dat er al een deel van de relatie afgebroken is. Die error kunnen we dan beter fixen dan een workaround te moeten maken. (fix hoeft niet in dit ticket)

Voor het ticket zelf: Verwijderen van een document zit meestal een bevestiging bij, want misklik is snel gebeurd en foetsie document.
Maar aangezien het niet in de story staat verwacht ik het ook niet. Kan later nog een vervolgstory worden.

Commit:
Merge branch 'development' into feature/KAS-2632-pub-proof--delete-piece

Commit:
KAS-2632 get subcase from controller

Commit:
KAS-2632 move model & pieceRow to route

Commit:
KAS-2632 remove subuse currentsession & subcase

Commit:
KAS-2632 delete piece in the right order

Commit:
Merge branch 'development' into feature/KAS-2632-pub-proof--delete-piece

TaneeG commented:
@ValenberghsSven Klaar voor verder review

ValenberghsSven committed:
KAS-2632 fix linter

ValenberghsSven reviewed (changes requested):
- in translations is `isDeleteDisabled` een alias voor `this.translationSubcase.isFinished` (`@action` gebruikt als getter).
In proofs skippen we de alias en gaan we direct voor `this.publicationSubcase.isFinished` in de template.

De alias maakt het beter leesbaar, ook al hebben we soms meerdere keer een gelijkaardige alias. (vb `isUploadDisabled` === `isDeleteDisabled` in translations/documents/controller)

- in translations heeft `isShownDelete` de `isOvrb` mee in de logica en is een `@action`, maar werkt als getter.
in proofs heeft `isShownDelete` die `isOvrb` logica niet maar in de template doen we die check wel, en dit is wel een getter.

de `isOvrb` toevoegingen kan ik niet zomaar zeggen wat daar de beste manier is.
Hangt ervan af wat we hiermee willen bereiken en welke profielen toegang hebben tot publicaties. 
Het was ook niet expliciet gevraagd in de story, laten we dat dan er voorlopig uit?
De delete afschermen ben ik sowieso voor, maar andere profielen (kabinet, minister, overheid) zien geen publicaties (geen propagatie) en hebben geen rechten om deletes uit te voeren.
De code nu zorgt er voor dat `kanselarij`geen deletes kan doen, maar ze kunnen wel andere acties uitvoeren (zoals vertalingen uploaden, aanvragen starten etc..).
Ik wil hier eerder aangeven dat het gebruik van `isOvrb` nu inconsistent is, en mogelijk doen we dit beter in een ander ticket

ValenberghsSven committed:
KAS-2632 never show the delete button if delete is not possible, made code more similar to each other

ValenberghsSven committed:
KAS-2632 add this. to template

ValenberghsSven committed:
Merge branch 'development' of github.com:kanselarij-vlaanderen/kaleidos-frontend into feature/KAS-2632-pub-proof--delete-piece

ValenberghsSven reviewed (approved):
Heb het zelf aangepast dus biased. Maar Lijkt me wel goed nu.

erikap commented on app/pods/publications/publication/proofs/documents/template.hbs:
```diff
@@ -79,66 +79,75 @@
+                    <Auk::Button
+                      @skin="borderless" @layout="icon-only" @icon="download"
+                      @href={{pieceRow.file.namedDownloadLink}}
+                      download={{await piece.downloadFilename}}>
```
Waarom de `await` hier? Kunnen we dan niet beter `pieceRow.file.downloadFilename` gebruiken?

erikap commented on app/pods/publications/publication/translations/documents/controller.js:
```diff
@@ -181,19 +188,17 @@ export default class PublicationsPublicationTranslationsDocumentController exten
-    return this.translationSubcase.isFinished
-      // can be translation or publication related
-      || piece.requestActivitiesUsedBy.length > 0;
+  isPieceDeletable(piece) {
```
Kunnen we in `translations/documents` dan ook niet beter met een `PieceRow` class werken zoals we nu in `proofs/documents` doen en deze actie als een get-function definiëren op die class? Dan werken de 2 routes analoog.

TaneeG replied:
@ValenberghsSven was hier van mening dat het misschien beter is m dit in een nieuw ticket te herwerken.
Als ik hem juist verstaan had.

ValenberghsSven replied:
Leek mij idd buiten scope + deze PR staat nu al redelijk lang open

erikap replied:
> de `isOvrb` toevoegingen kan ik niet zomaar zeggen wat daar de beste manier is.
> Hangt ervan af wat we hiermee willen bereiken en welke profielen toegang hebben tot publicaties.
> Het was ook niet expliciet gevraagd in de story, laten we dat dan er voorlopig uit?
> De delete afschermen ben ik sowieso voor, maar andere profielen (kabinet, minister, overheid) zien geen publicaties (geen propagatie) en hebben geen rechten om deletes uit te voeren.
> De code nu zorgt er voor dat `kanselarij`geen deletes kan doen, maar ze kunnen wel andere acties uitvoeren (zoals vertalingen uploaden, aanvragen starten etc..).
> Ik wil hier eerder aangeven dat het gebruik van `isOvrb` nu inconsistent is, en mogelijk doen we dit beter in een ander ticket

De `isOvrb`-check is woensdag besproken in de demo. Conclusie was dat iedereen die toegang heeft tot de publicaties ook alle acties moet kunnen doen. Er wordt nog nagevraagd wie allemaal toegang moet hebben tot de root-publications route.
Ik heb de bijkomende checks op `isOvrb` in de publications route/components eruit gehaald. Access tot de root-route wordt verder opgenomen in een vervolgstory.

Commit:
KAS-2632 await wegwerken

erikap committed:
Merge branch 'development' into feature/KAS-2632-pub-proof--delete-piece

erikap committed:
Remove unused await from setupController

erikap committed:
Remove additional checks on isOvrb in publication routes

As discussed in the demo the entire root-publications route will be
either accessible or unaccessible to the user. If it's accessible the
user will have permission to execute all actions. No additional checks
are required inside the publication routes/components.

erikap committed:
Apply pieceRow model on publication translations-documents route

Use same construction as in publication proofs-documents route

erikap committed:
Remove file property from PieceRow class for pub documents
